compiler/rustc_parse/src/parser/item.rs

```diff
@@ -374,25 +378,24 @@ impl<'a> Parser<'a> {
         self.check_ident() && self.look_ahead(1, |t| *t != token::Not && *t != token::PathSep)
     }
 
-    /// Recover on encountering a struct or method definition where the user
-    /// forgot to add the `struct` or `fn` keyword after writing `pub`: `pub S {}`.
+    /// Recover on encountering a struct, enum, or method definition where the user
+    /// forgot to add the `struct`, `enum`, or `fn` keyword
     fn recover_missing_kw_before_item(&mut self) -> PResult<'a, ()> {
-        // Space between `pub` keyword and the identifier
-        //
-        //     pub   S {}
-        //        ^^^ `sp` points here
-        let sp = self.prev_token.span.between(self.token.span);
-        let full_sp = self.prev_token.span.to(self.token.span);
+        let is_pub = self.prev_token.is_keyword(kw::Pub);
+        let is_const = self.prev_token.is_keyword(kw::Const);
+        let sp = self.token.span.shrink_to_lo();
```

Comment:
The span here is a bit suboptimal here. If there was a `pub`, we should use the span for `pub foo` (both tokens) in `pub foo() {}`. If there is no `pub`, I think it makes sense for us to just use the span for the identifier instead of shrinking.

As for the suggestion, I'm pretty sure you can use a separate (shrunk) span for the suggested snippet to replace.